### PR TITLE
Type-only auto imports

### LINKED ISSUE #36002: Auto import adds value import to type-only import statement
**TypeScript Version:**  3.8.0-dev.20200104

**Search Terms:**

**Code**

```ts
import type {SourceFile} from 'typescript';
const file: SourceFile = createSourceFile(
```

While typing `createSourceFile` use auto import.

**Expected behavior:**

Adds a named import for `createSourceFile` and removes the `type` keyword from the import.

**Actual behavior:**

Adds a named import for `createSourceFile` while keeping it a type-only import. This causes an error as `createSourceFile` does not exist in the type space. 

**Related Issues:** Introduced by #35200 @andrewbranch

### PR DESCRIPTION
**Pull request by andrewbranch**
Auto imports will not use type-only imports, _unless_ an import specifier can be added to an existing type-only import, and usage of the import occurs at a position where a type-only alias can be used:

```ts
import type { PathLike } from 'fs'

const dir: Dir/* | adds 'Dir' to existing type-only import */
```

On the other hand, if the import can be added to an existing type-only import, but is used in a position that is invalid for a type-only alias, the type-only import will be converted to a regular import:

```ts
import type { PathLike } from 'fs'

const dir = new Dir/* | adds 'Dir' to existing import and deletes 'type' */
```

GIF showing the two preceding code examples in action

From feedback in #36038, it sounds like we may also need to follow up and create a “hard mode” where an existing type-only import never converts to a regular import and blocks regular imports from that module from being added by auto-import. Such a hardline approach would probably frustrating, unexpected, and unnecessary for most users (as evidenced by the fact that #36002 was filed as a bug immediately) if enabled by default, but it’s justifiable to say “if I go through the trouble to write a type-only import, I _really_ want to make sure I don’t regular-import that module into this one.”

Screenshot of Minecraft “Hardcore Mode” enabled with setting description “World is deleted upon death.” I have crossed out and annotated it to read “Project is deleted upon value import.”

Fixes #36002

### CODE REVIEW

File: tests/cases/fourslash/importNameCodeFixNewImportTypeOnly.ts [hunk 1]
```diff
@@ -1,14 +0,0 @@
-/// <reference path="fourslash.ts" />
-
-// @target: esnext
-
-// @Filename: /a.ts
-////export interface Car {}
```
Comment (sheetalkamat) [inline tests/cases/fourslash/importNameCodeFixNewImportTypeOnly.ts]: Why was this deleted, is it because similar test already exists?

Comment (andrewbranch) [reply tests/cases/fourslash/importNameCodeFixNewImportTypeOnly.ts]: Yeah, it was actually verifying a behavior that we actually _don’t_ want anymore (using a type-only import for new imports of types), and reversing that behavior makes it redundant with many others.

**Comment by DimitryDushkin** [comment]
Hey! Could you please clarify how to enable auto-import with `type` for first import? AFAIK this pull can auto-import with `type` if such import already exists.

**Comment by andrewbranch** [comment]
An auto import will only be written as a type-only import if not doing so would result in an error. For that to be true, your tsconfig must have `importsNotUsedAsValues: "error"`, and the place where you’re performing the auto-import must be a type position.

**Comment by DimitryDushkin** [comment]
Thank you @andrewbranch !

**Comment by decryptingfuture** [comment]
> An auto import will only be written as a type-only import if not doing so would result in an error. For that to be true, your tsconfig must have `importsNotUsedAsValues: "error"`, and the place where you’re performing the auto-import must be a type position.

Awesome! It works, but.. the compiler gets stucked after such an error raises up and any change in a document does not invoke compiling. Am I the only one?

**Comment by Kos** [comment]
For those who find their way here from web search: The `importsNotUsedAsValues` tsconfig option is now deprecated, and the new one to look at is [`verbatimModuleSyntax`](https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax). I confirm vscode generates `import type` after enabling this flag.

**Comment by Hemaolle** [comment]
> For those who find their way here from web search: The `importsNotUsedAsValues` tsconfig option is now deprecated, and the new one to look at is [`verbatimModuleSyntax`](https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax). I confirm vscode generates `import type` after enabling this flag.

`verbatimModuleSyntax` was introduced in TS 5.0, so check your TS version before taking that option into use.
